File: chaos/loadgen.py
#!/usr/bin/env python3
"""Submits orders at a fixed rate and records every order id it was given.

Standard library only. Every logical order carries one Idempotency-Key; a
submit that gets no answer (connection refused, reset, timeout, 5xx) is
retried with the same key until the order service answers or the retry
budget is spent, so a kill of the order service costs latency, not orders.
Also polls the order service's stock endpoint so the synchronous inventory
path (breaker + time limiter) is exercised while the inventory service is
being killed.
"""
import argparse
import http.client
import json
import random
import threading
import time
import urllib.error
import urllib.request
import uuid

SKUS = ["SKU-ALPHA", "SKU-BRAVO", "SKU-CHARLIE", "SKU-SCARCE"]
WEIGHTS = [40, 30, 25, 5]
RETRY_STATUSES = {502, 503, 504}


def post_json(url, payload, key, timeout=5.0):
    body = json.dumps(payload).encode()
    headers = {"Content-Type": "application/json", "Idempotency-Key": key}
    req = urllib.request.Request(url, data=body, headers=headers)
    with urllib.request.urlopen(req, timeout=timeout) as resp:
        return json.loads(resp.read()), resp.headers.get("Idempotent-Replay", "false") == "true"


def get_json(url, timeout=3.0):
    with urllib.request.urlopen(url, timeout=timeout) as resp:
        return json.loads(resp.read())


def retryable(error):
    if isinstance(error, urllib.error.HTTPError):
        return error.code in RETRY_STATUSES
    return isinstance(error, (urllib.error.URLError, http.client.HTTPException, TimeoutError,
                              ConnectionError, OSError))


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--base", default="http://localhost:8081")
    parser.add_argument("--rate", type=float, default=20.0)
    parser.add_argument("--duration", type=float, default=60.0)
    parser.add_argument("--out", default="chaos/out/orders.json")
    parser.add_argument("--seed", type=int, default=7)
    parser.add_argument("--retry-seconds", type=float, default=90.0,
                        help="how long one order keeps retrying its key before it counts as an error")
    parser.add_argument("--retry-interval", type=float, default=1.0)
    args = parser.parse_args()

    rng = random.Random(args.seed)
    submitted, errors = [], []
    lock = threading.Lock()
    stop = threading.Event()
    stock_calls = {"live": 0, "cache": 0, "unknown": 0, "error": 0}
    counters = {"resubmits": 0, "retriedOrders": 0, "replayedAnswers": 0}

    def submit(seq, sku, qty, price):
        key = str(uuid.uuid4())
        payload = {
            "customerId": f"cust-{seq % 250}",
            "items": [{"sku": sku, "quantity": qty, "unitPrice": f"{price}.00"}],
        }
        t0 = time.time()
        attempts = 0
        while True:
            attempts += 1
            try:
                order, replayed = post_json(f"{args.base}/orders", payload, key)
            except Exception as e:  # noqa: BLE001
                if retryable(e) and time.time() - t0 < args.retry_seconds:
                    time.sleep(args.retry_interval)
                    continue
                with lock:
                    errors.append({"seq": seq, "key": key, "attempts": attempts,
                                   "error": str(e), "at": t0})
                return
            with lock:
                submitted.append({"id": order["id"], "key": key, "sku": sku, "quantity": qty,
                                  "submittedAt": t0, "acceptedAt": time.time(),
                                  "attempts": attempts, "replayed": replayed})
                if attempts > 1:
                    counters["resubmits"] += attempts - 1
                    counters["retriedOrders"] += 1
                if replayed:
                    counters["replayedAnswers"] += 1
            return

    def probe_stock():
        while not stop.is_set():
            try:
                view = get_json(f"{args.base}/stock/SKU-ALPHA")
                stock_calls[view.get("source", "unknown")] += 1
            except Exception:  # noqa: BLE001
                stock_calls["error"] += 1
            stop.wait(0.5)

    prober = threading.Thread(target=probe_stock, daemon=True)
    prober.start()

    interval = 1.0 / args.rate
    start = time.time()
    seq = 0
    threads = []
    while time.time() - start < args.duration:
        # the order mix is drawn on this thread so the sequence depends on the seed alone
        sku = rng.choices(SKUS, WEIGHTS)[0]
        qty = rng.randint(1, 3)
        price = rng.randint(5, 80)
        t = threading.Thread(target=submit, args=(seq, sku, qty, price), daemon=True)
        t.start()
        threads.append(t)
        seq += 1
        next_at = start + seq * interval
        delay = next_at - time.time()
        if delay > 0:
            time.sleep(delay)
    deadline = time.time() + args.retry_seconds + 15
    for t in threads:
        t.join(timeout=max(0.0, deadline - time.time()))
    stop.set()
    prober.join(timeout=2)

    result = {
        "rate": args.rate,
        "duration": args.duration,
        "seed": args.seed,
        "retrySeconds": args.retry_seconds,
        "startedAt": start,
        "finishedAt": time.time(),
        "submitted": submitted,
        "submitErrors": errors,
        "stockProbes": stock_calls,
        **counters,
    }
    with open(args.out, "w") as fh:
        json.dump(result, fh)
    print(f"loadgen: {len(submitted)} orders submitted, {len(errors)} submit errors, "
          f"{counters['resubmits']} resubmits over {counters['retriedOrders']} orders, "
          f"{counters['replayedAnswers']} answered from the idempotency store, "
          f"stock probes {stock_calls}")


if __name__ == "__main__":
    main()

File: web/src/sim/loadgen.ts
import type { Cluster } from "./cluster.ts";
import type { ServiceName } from "./events.ts";
import { Prng } from "./prng.ts";

export const SKUS = ["SKU-ALPHA", "SKU-BRAVO", "SKU-CHARLIE", "SKU-SCARCE"] as const;
const WEIGHTS = [40, 30, 25, 5];

export interface LoadOptions {
  rate: number;
  durationMs: number;
  seed?: number;
  probeMs?: number;
}

/**
 * Mirrors chaos/loadgen.py: a fixed order rate with weighted skus, quantities 1..3, prices 5..80,
 * and a stock probe every 500 ms against the order service's breaker protected endpoint.
 */
export class LoadGenerator {
  private readonly rng: Prng;
  private startedAt = -1;
  private seq = 0;
  private nextProbeAt = 0;
  readonly probeSku = "SKU-ALPHA";

  constructor(private readonly options: LoadOptions) {
    this.rng = new Prng(options.seed ?? 7);
  }

  get elapsed(): number {
    return this.startedAt < 0 ? 0 : Math.max(0, this.elapsedFrom);
  }

  private elapsedFrom = 0;

  finished(now: number): boolean {
    return this.startedAt >= 0 && now - this.startedAt >= this.options.durationMs;
  }

  /** Submit whatever is due at `cluster.now`. Returns how many orders were submitted. */
  tick(cluster: Cluster): number {
    const now = cluster.now;
    if (this.startedAt < 0) this.startedAt = now;
    this.elapsedFrom = now - this.startedAt;
    let count = 0;
    const interval = 1000 / this.options.rate;
    while (
      now - this.startedAt < this.options.durationMs &&
      this.startedAt + this.seq * interval <= now
    ) {
      const sku = this.rng.choice(SKUS, WEIGHTS);
      const quantity = this.rng.int(1, 3);
      const unitPrice = this.rng.int(5, 80);
      cluster.submit(`cust-${this.seq % 250}`, [{ sku, quantity, unitPrice }]);
      this.seq++;
      count++;
    }
    if (now >= this.nextProbeAt && now - this.startedAt < this.options.durationMs) {
      this.nextProbeAt = now + (this.options.probeMs ?? 500);
      cluster.probeStock(this.probeSku);
    }
    return count;
  }
}

export interface KillPlan {
  at: number;
  service: ServiceName;
}

/** The kill schedule from chaos/run.sh: kills spread over the window with jitter. */
export function planKills(durationMs: number, kills: number, seed: number): KillPlan[] {
  const rng = new Prng(seed ^ 0x5eed);
  const victims: ServiceName[] = ["inventory-service", "payment-service"];
  const slot = Math.floor(durationMs / 1000 / (kills + 1));
  const plan: KillPlan[] = [];
  for (let i = 0; i < kills; i++) {
    let target = slot * (i + 1) + rng.int(0, Math.floor(slot / 2)) - Math.floor(slot / 4);
    if (target < 5) target = 5;
    plan.push({ at: target * 1000, service: victims[rng.int(0, victims.length - 1)] });
  }
  return plan;
}
